src/core/collector/mod.rs

```diff
@@ -0,0 +1,757 @@
+	listen_addr: Option<SocketAddr>,
+}
+
+/// This type is used to distinguish different keys in the storage
+#[derive(Clone, Copy, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum CollectorPrefixType {
+	/// Candidate prefixed by Parachain-Id
+	Candidate(u32),
+	/// A mapping to find out parachain id by candidate hash (e.g. when we don't know the parachain)
+	CandidatesParachains,
+	/// Relay chain block
+	Head,
+	/// Session data, keyed by session hash (blake2b(session_index))
+	Session,
+	/// Inherent data (more expensive to store, so good to have it shared)
+	InherentData,
+	/// Dispute information indexed by Parachain-Id; data is DisputeInfo
+	Dispute(u32),
+}
+
+/// A type that defines prefix + hash itself
+pub(crate) type CollectorStorageApi = ApiService<H256, CollectorPrefixType>;
+
+/// A structure used to track disputes progress
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct DisputeInfo {
+	pub initiated: <PolkadotConfig as subxt::Config>::BlockNumber,
+	pub dispute: SubxtDispute,
+	pub parachain_id: u32,
+	pub outcome: Option<SubxtDisputeResult>,
+	pub concluded: Option<<PolkadotConfig as subxt::Config>::BlockNumber>,
+}
+
+/// The current state of the collector, used to detect forks and track candidates
+/// and other events without query storage
+#[derive(Default)]
+struct CollectorState {
+	/// The current block number
+	current_relay_chain_block_number: u32,
+	/// A list of hashes at this height (e.g. if we see forks)
+	current_relay_chain_block_hashes: Vec<H256>,
+	/// A list of candidates seen, indexed by parachain id
+	candidates_seen: BTreeMap<u32, Vec<H256>>,
+	/// A list of disputes seen, indexed by parachain id
+	disputes_seen: BTreeMap<u32, Vec<DisputeInfo>>,
+	/// A current session index
+	current_session_index: u32,
+}
+
+/// Provides collector new head events split by parachain
+#[derive(Clone, Debug)]
+pub struct NewHeadEvent {
+	/// Relay parent block number
+	pub relay_parent_number: <PolkadotConfig as subxt::Config>::BlockNumber,
+	/// Relay parent block hash (or hashes in case of the forks)
+	pub relay_parent_hashes: Vec<H256>,
+	/// The target parachain id (used for broadcasting events)
+	pub para_id: u32,
+	/// Candidates seen for this relay chain block that belong to the target parachain
+	pub candidates_seen: Vec<H256>,
+	/// Disputes concluded in these blocks
+	pub disputes_concluded: Vec<DisputeInfo>,
+}
+
+/// Handles collector updates
+#[derive(Clone, Debug)]
+pub enum CollectorUpdateEvent {
+	/// Occurs on new block processing with the information about the previous block
+	NewHead(NewHeadEvent),
+	/// Occurs on a new session
+	NewSession(u32),
+	/// Occurs when collector is disconnected and is about to terminate
+	Termination,
+}
+
+pub struct Collector {
+	api_service: CollectorStorageApi,
+	ws_listener: Option<WebSocketListener>,
+	to_websocket: Option<BroadcastSender<WebSocketUpdateEvent>>,
+	endpoint: String,
+	subscribe_channels: BTreeMap<u32, Vec<BroadcastSender<CollectorUpdateEvent>>>,
```

Comment:
I am wondering why we need both subscriptions and broadcast channels, for simplicity we could only use broadcast and filter by para id at consumer level.

Reply:
Mainly for the convenience. Per-parachain subscriptions are the only subscriptions that exist for now. We might want broadcast subscriptions at some point as well. Filtering on the consumer side is possible but might be noisy, as we normally need a single parachain but will receive quite a lot of traffic about unrelated events.